README.md
# `remark-lint-frontmatter-schema` 📑

<!-- [![Build Status](https://img.shields.io/github/workflow/status/JulianCataldo/remark-lint-frontmatter-schema/release/master.svg)](https://github.com/remark-lint-frontmatter-schema/actions/workflows/release.yml?query=branch%3Amain) -->

[![VS Code](https://img.shields.io/badge/Visual_Studio_Code-0078D4?logo=visual%20studio%20code)](https://code.visualstudio.com)
[![unified](https://img.shields.io/badge/uni-fied-0366d6?logo=markdown)](https://unifiedjs.com)  
[![NPM](https://img.shields.io/npm/v/remark-lint-frontmatter-schema)](https://www.npmjs.com/package/remark-lint-frontmatter-schema)
![Downloads](https://img.shields.io/npm/dt/remark-lint-frontmatter-schema)
[![ISC License](https://img.shields.io/npm/l/remark-lint-frontmatter-schema)](./LICENSE)
[![GitHub](https://img.shields.io/badge/Repository-222222?logo=github)](https://github.com/JulianCataldo/remark-lint-frontmatter-schema)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen)](https://makeapullrequest.com)  
[![TypeScript](https://img.shields.io/badge/TypeScript-333333?logo=typescript)](http://www.typescriptlang.org/)
[![Prettier](https://img.shields.io/badge/Prettier-333333?logo=prettier)](https://prettier.io)
[![EditorConfig](https://img.shields.io/badge/EditorConfig-333333?logo=editorconfig)](https://editorconfig.org)
[![ESLint](https://img.shields.io/badge/ESLint-3A33D1?logo=eslint)](https://eslint.org)

<!-- [![Renovate](https://img.shields.io/badge/Renovate-enabled-17a2b8?logo=renovatebot)](https://app.renovatebot.com/dashboard) -->

Validate **Markdown** frontmatter **YAML** against an associated **JSON schema** with this **remark-lint** rule plugin.

Supports:

- **Types validation**, pattern, enumerations,… and all you can get with JSON Schema
- **Code location** problems indicator (for IDE to underline)
- **Auto-fixes** with suggestions
- **C**ommand **L**ine **I**nterface reports
- **VS Code** integration (see below)
- **Global patterns** or **in-file** schemas associations
- In JS framework **MD / MDX pipelines**

# Demo

<div align="center">

[![](https://res.cloudinary.com/dzfylx93l/image/upload/c_scale,w_1280/demo-rlfmschema_meai5w.png)  
**🕹  Preview it online!**](https://astro-content.dev/__content)

<sup><sub>(w. Astro Content — Editor)</sub></sup>

</div>

---

**Jump to**:

- [👉  **Play with pre-configured ./demo**](#play-with-pre-configured-demo)
  - [Base](#base)
  - [VS Code (optional)](#vs-code-optional)
  - [CLI / IDE (VS Code) — **Static** linting](#cli--ide-vs-code--static-linting)
    - [Workspace](#workspace)
    - [Schema example](#schema-example)
      - [🆕  Add references to external definitions (advanced)](#add-references-to-external-definitions-advanced)
    - [Schemas associations](#schemas-associations)
      - [Inside frontmatter](#inside-frontmatter)
      - [Globally, with patterns](#globally-with-patterns)
    - [CLI usage](#cli-usage)
    - [Bonus](#bonus)
      - [Validate your schema with _JSON meta schema_](#validate-your-schema-with-json-meta-schema)
      - [ESLint MDX plugin setup](#eslint-mdx-plugin-setup)
        - [Known issues](#known-issues)
  - [MD / MDX pipeline — **Runtime** validation](#md--mdx-pipeline--runtime-validation)
    - [Custom pipeline](#custom-pipeline)
      - [Implementation living example](#implementation-living-example)
      - [Important foot-notes for custom pipeline](#important-foot-notes-for-custom-pipeline)
    - [Framework](#framework)
      - [Astro](#astro)
      - [Gatsby](#gatsby)

---

[![Demo screenshot of frontmatter schema linter 1](./docs/screenshot.png)](https://raw.githubusercontent.com/JulianCataldo/remark-lint-frontmatter-schema/master/docs/screenshot.png)

---

[![Demo screenshot of frontmatter schema linter 2](./docs/screenshot-2.png)](https://raw.githubusercontent.com/JulianCataldo/remark-lint-frontmatter-schema/master/docs/screenshot-2.png)

---

[![Demo screenshot of frontmatter schema linter 3](./docs/screenshot-3.png)](https://raw.githubusercontent.com/JulianCataldo/remark-lint-frontmatter-schema/master/docs/screenshot-3.png)

---

## 👉  **Play with pre-configured [./demo](./demo/)**

Quick shallow **clone** with:

```sh
pnpx degit JulianCataldo/remark-lint-frontmatter-schema/demo ./demo
```

---

# Installation

### Base

```sh
pnpm install -D \
remark remark-cli \
remark-frontmatter \
remark-lint-frontmatter-schema
```

> **Remove** `-D` flag for runtime **`unified`** MD / MDX **pipeline** (custom, Astro, Gatsby, etc.), for production.  
> **Keep it** if you just want to lint with **CLI** or your **IDE** locally, without any production / CI needs.

### VS Code (optional)

```sh
code --install-extension unifiedjs.vscode-remark
```

# Configuration

### CLI / IDE (VS Code) — **Static** linting

👉  **See [./demo](./demo/)** folder to get a working, pre-configured, bare project workspace.  
You also get example Markdown files and associated schema to play with.  
Supports `remark-cli` and/or `unifiedjs.vscode-remark` extension.

📌  Check out the **[demo/README.md](./demo) for bootstrapping** it.

#### Workspace

Create the root config. file for `remark` to source from:  
`touch ./.remarkrc.mjs`

Paste this base configuration:

```mjs
import remarkFrontmatter from 'remark-frontmatter';
import remarkLintFrontmatterSchema from 'remark-lint-frontmatter-schema';

const remarkConfig = {
	plugins: [remarkFrontmatter, remarkLintFrontmatterSchema],
};
export default remarkConfig;
```

You can use YAML / JSON / …, too (uses [cosmiconfig](https://github.com/davidtheclark/cosmiconfig)).

#### Schema example

`./content/creative-work.schema.yaml`

```yaml
type: object
properties:
  title:
    type: string
# …
```

##### 🆕  Add references to external definitions (advanced)

Referencing schema definitions
allows re-using bit and piece instead of duplicate them,
accross your content schemas.

You can reference an external schema relatively, using `$ref`.
For example we can -_kind of_- merge an host object with a reference properties:

The host schema, `content/articles/index.schema.yaml`

```yaml
allOf:
  - $ref: ../page.schema.yaml

  - properties:
      layout:
        const: src/layouts/Article.astro
      category:
        type: string
        enum:
          - Book
          - Movie
      foo:
        type: string

    required:
      - layout
      - category
```

A referenced schema, `content/page.schema.yaml`

```yaml
properties:
  title:
    type: string
    maxLength: 80
    # ...
  # ...

required:
  - title
```

The result will be _(virtually)_ the same as this:

```yaml
properties:
  title:
    type: string
    maxLength: 80
    # ...
  # ...
  layout:
    const: src/layouts/Article.astro
  category:
    type: string
    enum:
      - Book
      - Movie
  foo:
    type: string
  # ...

required:
  - title
  - layout
  - category
```

#### Schemas associations

Inspired by [VS Code JSON Schema](https://code.visualstudio.com/docs/languages/json#_json-schemas-and-settings)
and [`redhat.vscode-yaml`](https://marketplace.visualstudio.com/items?itemName=redhat.vscode-yaml) conventions.

##### Inside frontmatter

See **[./demo/content](./demo/content)** files for examples.

Schema association can be done directly **inside** the **frontmatter** of the **Markdown** file,
relative to project root, thanks to the `'$schema'` key:

```markdown
---
# From workspace root (`foo/…`, `/foo/…` or `./foo/…` is the same)
'$schema': content/creative-work.schema.yaml

# —Or— relatively, from this current file directory (`./foo/…` or `../foo/…`)
# '$schema': ../creative-work.schema.yaml

layout: src/layouts/Article.astro

title: Hello there
category: Book
# …
---

# You're welcome!

🌝  My **Markdown** content…  🌚
…
```

##### Globally, with patterns

> **Note**:  
> Locally defined **`'$schema'` takes precedence** over global settings below.

```js
const remarkConfig = {
	plugins: [
		remarkFrontmatter,
		[
			remarkLintFrontmatterSchema,
			{
				schemas: {
					/* One schema for many files */
					'./content/creative-work.schema.yaml': [
						/* Per-file association */
						'./content/creative-work/the-shipwreck__global-broken.md',

						/* Support glob patterns ———v */
						// './content/creative-work/*.md',
						// …
						// `./` prefix is optional
						// 'content/creative-work/foobiz.md',
					],

					// './content/ghost.schema.yaml': [
					//   './content/casper.md',
					//   './content/ether.md',
					// ],
				},
			},
		],
	],
};
```

`'./foo'`, `'/foo'`, `'foo'`, all will work.  
It's always relative to your `./.remarkrc.mjs` file, in your workspace root.

#### CLI usage

Linting whole workspace files (as `./**/*.md`) with `remark-cli`:

```sh
pnpm remark .
```

Yields:

![](https://res.cloudinary.com/dzfylx93l/image/upload/v1666912219/Xnapper-2022-10-28-01.09.11_yh4tnr.png)

#### Bonus

##### Validate your schema with _JSON meta schema_

First, install the [YAML for Visual Studio Code](https://marketplace.visualstudio.com/items?itemName=redhat.vscode-yaml) extension:

```sh
code --install-extension redhat.vscode-yaml
```

Then, add this to your `.vscode/settings.json`:

```jsonc
{
	"yaml.schemas": {
		"http://json-schema.org/draft-07/schema#": ["content/**/*.schema.yaml"]
	}
	/* ... */
}
```

##### ESLint MDX plugin setup

Will work with the ESLint VS Code extension and the CLI command.

Install the [ESLint MDX plugin](https://github.com/mdx-js/eslint-mdx),
the [MDX VS Code extension](https://marketplace.visualstudio.com/items?itemName=unifiedjs.vscode-mdx) and the [ESLint VS Code extension](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint).

Add this dependencies to your project:

```sh
pnpm i -D eslint eslint-plugin-mdx \
eslint-plugin-prettier eslint-config-prettier
```

Add a `.eslintrc.cjs`:

```js
/** @type {import("@types/eslint").Linter.Config} */

module.exports = {
	overrides: [
		{
			files: ['*.md', '*.mdx'],
			extends: ['plugin:mdx/recommended'],
		},
	],
};
```

Add a `.remarkrc.yaml` (or JSON, etc.), e.g.:

```yaml
plugins:
  - remark-frontmatter

  - - remark-lint-frontmatter-schema
    - schemas:
        src/schemas/blog-post.schema.yaml:
          - content/blog-posts/*.{md,mdx}

  # - remark-preset-lint-consistent
  # - remark-preset-lint-markdown-style-guide
  # - remark-preset-lint-recommended
```

---

Result:

[![](https://res.cloudinary.com/dzfylx93l/image/upload/c_scale,w_1280/eslint-plugin-mdx-1.png)  
](https://res.cloudinary.com/dzfylx93l/image/upload/eslint-plugin-mdx-1.png)

---

Lint with CLI:

```sh
pnpm eslint --ext .mdx .
```

> Efforts has been made to have the best output for both remark and ESLint,
> for IDE extensions and CLIs.

###### Known issues

- Expected `enum` values suggestions are working with the remark extension, not with the ESLint one.
- Similarly, ESLint output will give less details (see screenshot above), and a bit different layout for CLI output, too.
- remark extension seems to load faster, and is more reactive to schema changes.
- As of `eslint-plugin-mdx@2`, `.remarkrc.mjs` (ES Module) is not loaded, JSON and YAML configs are fine.

### MD / MDX pipeline — **Runtime** validation

Use it as usual like any remark plugin inside your framework or your custom `unified` pipeline.

#### Custom pipeline

When processing Markdown as single files inside your JS/TS app.  
An minimal example is provided in [`./demo/pipeline.ts`](./demo/pipeline.ts), you can launch it with `pnpm pipeline` from `./demo`.

---

Schema should be provided programmatically like this:

```ts
// …
import remarkFrontmatter from 'remark-frontmatter';
import remarkLintFrontmatterSchema from 'remark-lint-frontmatter-schema';
import type { JSONSchema7 } from 'json-schema';
import { reporter } from 'vfile-reporter';

const mySchema: JSONSchema7 = {
	/* … */
};

const output = await unified()
	// Your pipeline (basic example)
	.use(remarkParse)
	// …
	.use(remarkFrontmatter)

	.use(remarkLintFrontmatterSchema, {
		/* Bring your own schema */
		embed: mySchema,
	})

	// …
	.use(remarkRehype)
	.use(rehypeStringify)
	.use(rehypeFormat)
	.process(theRawMarkdownLiteral);

/* `path` is for debugging purpose here, as MD literal comes from your app. */
output.path = './the-current-processed-md-file.md';

console.error(reporter([output]));
```

Yields:

```
./the-current-processed-md-file.md
  1:1  warning  Must have required property 'tag'  frontmatter-schema  remark-lint

⚠ 1 warning
```

##### Implementation living example

Checkout [**Astro Content**](https://github.com/JulianCataldo/astro-content) repository.

<!-- It's a text based, structured content assistant, integrated in Astro, for edition and consumption.   -->
<!-- file or API based  -->

Astro Content relies on this library, among others, for providing linting reports.

<!-- You can see **remark-lint-frontmatter-schema** in action, on **[this line, in Astro Content source](outdated)**. -->

##### Important foot-notes for custom pipeline

This is **different from static linting**, with VS Code extension or CLI.  
It **will not source `.remarkrc`** (but you can source it by your own means, if you want).  
In fact, it's not aware of your file structure,
nor it will associate or import any schema / Markdown files.  
That way, it will integrate easier with your own business logic and existing pipelines.  
I found that **static linting** (during editing) / and **runtime validation** are two different
uses cases enough to separate them in their setups, but I might converge them partially.

#### Framework

> **Warning**  
> WIP. **NOT tested yet**! It is not a common use case for `remark-lint`.  
> Linting data inside frameworks are generally ignored.  
> AFAIK, `messages` data isn't forwarded to CLI output.  
> Feel free to open a PR if you have some uses cases in this area that need special care.  
> Maybe Astro or Astro Content could leverage these linter warnings in the future.

See [global patterns `schemas` associations](#globally-with-patterns) for settings reference.

##### Astro

In `astro.config.mjs`

```ts
// …
export default defineConfig({
  // …
  remarkPlugins: [
    // …
    'remark-frontmatter',
    ['remark-lint-frontmatter-schema', { schemas }],
    // …
  ];
  // …
});
```

##### Gatsby

In `gatsby-config.js`

```ts
{
	// …
	plugins: [
		// …
		{
			resolve: 'gatsby-transformer-remark',
			options: {
				plugins: [
					// …
					'remark-frontmatter',
					['remark-lint-frontmatter-schema', { schemas }],
					// …
				],
			},
		},
		// …
	];
}
```

# Interfaces

```ts
export interface Settings {
	/**
	 * Global workspace file associations mapping (for linter extension).
	 *
	 * **Example**: `'schemas/thing.schema.yaml': ['content/things/*.md']`
	 */
	schemas?: Record<string, string[]>;

	/**
	 * Direct schema embedding (for using inside an `unified` transform pipeline).
	 *
	 * Format: JSON Schema - draft-2019-09
	 *
	 * **Documentation**: https://ajv.js.org/json-schema.html#draft-07
	 */
	embed?: JSONSchema7;

	/**
	 * **Documentation**: https://ajv.js.org/options.html
	 */
	ajvOptions?: AjvOptions;
}


CHANGELOG.md
## [3.15.4](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/compare/v3.15.3...v3.15.4) (2023-10-15)


### Bug Fixes

* upgrade deps, fix demo ([9dd9dbb](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/commit/9dd9dbb896fbe7a2c63c7651a8cc3899e0b3ec65))

## [3.15.3](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/compare/v3.15.2...v3.15.3) (2023-09-15)


### Bug Fixes

* treat first yaml section as frontmatter, even if it is not the first child ([c09c902](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/commit/c09c9021fc274033824be9728ffdcd3023dd3619))

## [3.15.2](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/compare/v3.15.1...v3.15.2) (2023-04-12)


### Bug Fixes

* trigger release ([11db24b](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/commit/11db24b9fb8851e28417e42e912a43617279cddb))

## [3.15.1](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/compare/v3.15.0...v3.15.1) (2022-11-19)


### Bug Fixes

* better configuration example (yaml) ([c4629bb](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/commit/c4629bb926d6f8be92bf7a580188874440dd783f))

# [3.15.0](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/compare/v3.14.0...v3.15.0) (2022-11-03)


### Features

* bump major version ([336f10e](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/commit/336f10e4c37819e6cd1a6ff9e490d72e26e578c1))

# [3.14.0](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/compare/v3.13.0...v3.14.0) (2022-11-03)


### Bug Fixes

* align demo with new feats ([3cbc16a](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/commit/3cbc16a20c3ceae772548f322d949ce978550b8e))
* use bundle instead of dereference ([6a8824a](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/commit/6a8824a521015ec083408694e2101b914140efb8))


### Features

* resolve current remark working directory ([1887ef1](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/commit/1887ef1e857b2deaf64fd542ffca1f328d298628))
* use only json-schema-ref-parser, drop ajv's ([99b66bd](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/commit/99b66bd7fb36d8e17daf001976ab41eb6b133ecd))

# [3.13.0](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/compare/v3.12.0...v3.13.0) (2022-10-30)


### Features

* support relative path for local `$schema` ([20f53bf](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/commit/20f53bff49512e27a0b04a304cead2cecd0859b8))

# [3.12.0](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/compare/v3.11.0...v3.12.0) (2022-10-30)


### Features

* support for `const` auto-fix / suggestion ([a133c54](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/commit/a133c54eeb322c03e68582d80036fa39194d161e))

# [3.11.0](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/compare/v3.10.2...v3.11.0) (2022-10-30)


### Features

* better error output for eslint / remark ([aff95db](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/commit/aff95db261f45eb4dd8f1a947640ec0a312c8772))

## [3.10.2](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/compare/v3.10.1...v3.10.2) (2022-10-29)


### Bug Fixes

* global schema assoc. w. `eslint-plugin-mdx` ([6849182](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/commit/6849182449db52e7cd1ea76b671466e3c690cd9c))

## [3.10.1](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/compare/v3.10.0...v3.10.1) (2022-10-29)


### Bug Fixes

* allow empty settings for runtime plugin ([7534ea4](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/commit/7534ea4287abfa772779944839311acc7dcdd354))

# [3.10.0](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/compare/v3.9.0...v3.10.0) (2022-10-29)


### Features

* embedded `$ref` for demo pipeline ([59434d7](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/commit/59434d78470e4e0d5fc0756ddf988f40acd862e5))

# [3.9.0](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/compare/v3.8.0...v3.9.0) (2022-10-29)


### Features

* convert lint rule to async, for file loading ([4675b25](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/commit/4675b256fcd84a703deeebf3b412d2023c0cac6e))
* load external schema definition references ([b1b9805](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/commit/b1b98051d03c17b6cc2bc3a29ebb8206172a52c0))

# [3.8.0](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/compare/v3.7.9...v3.8.0) (2022-10-28)


### Features

* core meta schema validation docs + example ([1976cc2](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/commit/1976cc29f486448fbd72db45f0b9156214e0d430))
* embed native errors reports in vfile msg ([5a39364](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/commit/5a393647d61f7e2730f07649891c2b0347729ced))

## [3.7.9](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/compare/v3.7.8...v3.7.9) (2022-10-27)


### Bug Fixes

* new cli harvest image + remove pointing hand ([42c5658](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/commit/42c56581598f919b0060f315196c639b8d18b8b9))

## [3.7.8](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/compare/v3.7.7...v3.7.8) (2022-10-27)


### Bug Fixes

* update badge, links,… + pjson ordering ([8ebaca9](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/commit/8ebaca9b11cb3ec97e13e2ee84adaff5cb43df7e))

## [3.7.7](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/compare/v3.7.6...v3.7.7) (2022-10-20)


### Bug Fixes

* properties spacing + bold ([e168cab](https://github.com/JulianCataldo/remark-lint-frontmatter-schema/commit/e168cabd2d2cbf7619f07da43efd506d54386dae))


index.ts
/* ——————————————————————————————————————————————————————————————————————————— *
 *              © Julian Cataldo — https://www.juliancataldo.com.              *
 *                      See LICENSE in the project root.                       *
/* —————————————————————————————————————————————————————————————————————————— */

/* eslint-disable max-lines */
import path from 'node:path';
import { existsSync } from 'node:fs';
import { findUp } from 'find-up';
// NOTE: minimatch@9 is breaking import.
import { minimatch } from 'minimatch';
/* ·········································································· */
import yaml, { type Document, isNode, LineCounter } from 'yaml';
import Ajv from 'ajv';
import type { Options as AjvOptions, ErrorObject as AjvErrorObject } from 'ajv';
import addFormats from 'ajv-formats';
import $RefParser from '@apidevtools/json-schema-ref-parser';
import type { JSONSchema7 } from 'json-schema';
/* ·········································································· */
import { lintRule } from 'unified-lint-rule';
import type { VFile } from 'unified-lint-rule/lib';
import type { Root, YAML } from 'mdast';
import type { VFileMessage } from 'vfile-message';
/* —————————————————————————————————————————————————————————————————————————— */

const url = 'https://github.com/JulianCataldo/remark-lint-frontmatter-schema';
const nativeJsErrorName = 'Markdown YAML frontmatter error (JSON Schema)';

export interface Settings {
	/**
	 * Global workspace file associations mapping (for linter extension).
	 *
	 * **Example**: `'schemas/thing.schema.yaml': ['content/things/*.md']`
	 */
	schemas?: Record<string, string[]>;

	/**
	 * Direct schema embedding (for using inside an `unified` transform pipeline).
	 *
	 * Format: JSON Schema - draft-2019-09
	 *
	 * **Documentation**: https://ajv.js.org/json-schema.html#draft-07
	 */
	embed?: JSONSchema7;

	/**
	 * **Documentation**: https://ajv.js.org/options.html
	 */
	ajvOptions?: AjvOptions;
}

// IDEA: Might be interesting to populate with corresponding error reference
type JSONSchemaReference = 'https://ajv.js.org/json-schema.html';

export interface FrontmatterSchemaMessage extends VFileMessage {
	schema: AjvErrorObject & { url: JSONSchemaReference };
}

interface FrontmatterObject {
	$schema: string | undefined;
	/* This is the typical Frontmatter object, as treated by common consumers */
	[key: string]: unknown;
}

/* ·········································································· */

/* The vFile cwd isn't the same as the one from IDE extension.
Extension will cascade upward from the current processed file and
take the remarkrc file as its cwd. It's multi-level workspace
friendly. We have to mimick this behavior here, as remark lint rules doesn't 
seems to offer an API to hook up on this? */
async function getRemarkCwd(startDir: string) {
	const remarkConfigNames = [
		'.remarkrc',
		'.remarkrc.json',
		'.remarkrc.yaml',
		'.remarkrc.yml',
		'.remarkrc.mjs',
		'.remarkrc.js',
		'.remarkrc.cjs',
	];
	const remarkConfPath = await findUp(remarkConfigNames, {
		cwd: path.dirname(startDir),
	});
	if (remarkConfPath) {
		return path.dirname(remarkConfPath);
	}
	return process.cwd();
}

/* ·········································································· */

function pushErrors(
	errors: AjvErrorObject[],
	yamlDoc: Document,
	vFile: VFile,
	/** File path from local `$schema` key or from global settings */
	schemaRelPath: string,
	/** Used to map character range to line / column tuple */
	lineCounter: LineCounter,
) {
	errors.forEach((error) => {
		let reason = '';

		if (error.message) {
			/* Capitalize error message */
			const errMessage =
				`${error.message.charAt(0).toUpperCase()}` +
				`${error.message.substring(1)}`;

			let expected = '';
			if (Array.isArray(error.params.allowedValues)) {
				expected = `: \`${error.params.allowedValues.join('`, `')}\``;
			} else if (typeof error.params.allowedValue === 'string') {
				expected = `: \`${error.params.allowedValue}\``;
			}
			const sPath = schemaRelPath ? ` • ${schemaRelPath}` : '';

			reason = `${errMessage}${expected}${sPath} • ${error.schemaPath}`;
		}

		/* Explode AJV error instance path and get corresponding YAML AST node */
		const ajvPath = error.instancePath.substring(1).split('/');
		const node = yamlDoc.getIn(ajvPath, true);

		const message = vFile.message(reason);

		// FIXME: Doesn't seems to be used in custom pipeline?
		// Always returning `false`
		message.fatal = true;

		/* `name` comes from native JS `Error` object */
		message.name = nativeJsErrorName;

		/* Map YAML characters range to column / line positions,
      -OR- squiggling the opening frontmatter fence for **root** path errors */
		if (isNode(node)) {
			/* Incriminated token */
			message.actual = node.toString();

			/* Map AJV Range to VFile Position, via YAML lib. parser */
			if (node.range) {
				const OPENING_FENCE_LINE_COUNT = 1; /* Takes the `---` into account */
				const start = lineCounter.linePos(node.range[0]);
				const end = lineCounter.linePos(node.range[1]);
				message.position = {
					start: {
						line: start.line + OPENING_FENCE_LINE_COUNT,
						column: start.col,
					},
					end: {
						line: end.line + OPENING_FENCE_LINE_COUNT,
						column: end.col,
					},
				};
				// NOTE: Seems redundant, but otherwise, it is always set to 1:1 */
				message.line = message.position.start.line;
				message.column = message.position.start.column;
			}
		}

		/* Assemble pretty per-error insights for end-user */
		let note = `Keyword: ${error.keyword}`;
		if (Array.isArray(error.params.allowedValues)) {
			note += `\nAllowed values: ${error.params.allowedValues.join(', ')}`;

			/* Auto-fix replacement suggestions for `enum` */
			message.expected = error.params.allowedValues;
		} else if (typeof error.params.allowedValue === 'string') {
			note += `\nAllowed value: ${error.params.allowedValue}`;

			/* Auto-fix replacement suggestion for `const` */
			message.expected = [error.params.allowedValue];
		}
		if (typeof error.params.missingProperty === 'string') {
			note += `\nMissing property: ${error.params.missingProperty}`;
		}
		if (typeof error.params.type === 'string') {
			note += `\nType: ${error.params.type}`;
		}
		/* `schemaRelPath` path prefix will show up only when using
      file association, not when using pipeline embedded schema */
		note += `\nSchema path: ${schemaRelPath} · ${error.schemaPath}`;
		message.note = note;
		/* `message` comes from native JS `Error` object */
		message.message = reason;

		/* Adding custom data from AJV */
		/* It’s OK to store custom data directly on the VFileMessage:
      https://github.com/vfile/vfile-message#well-known-fields */
		// NOTE: Might be better to type `message` before, instead of asserting here
		(message as FrontmatterSchemaMessage).schema = {
			url: 'https://ajv.js.org/json-schema.html',
			...error,
		};
	});
}

/* ·········································································· */

async function validateFrontmatter(
	sourceYaml: YAML,
	vFile: VFile,
	settings: Settings,
) {
	const hasPropSchema = typeof settings.embed === 'object';
	const lineCounter = new LineCounter();
	let yamlDoc;
	let yamlJS;
	let hasLocalAssoc = false;
	let schemaPathFromCwd: string | undefined;
	const remarkCwd = await getRemarkCwd(vFile.path);

	/* Parse the YAML literal and get the YAML Abstract Syntax Tree,
    previously extracted by `remark-frontmatter` */
	try {
		yamlDoc = yaml.parseDocument(sourceYaml.value, { lineCounter });
		yamlJS = yamlDoc.toJS() as FrontmatterObject | null;

		/* Local `$schema` association takes precedence over global / prop. */
		if (yamlJS?.$schema && typeof yamlJS.$schema === 'string') {
			hasLocalAssoc = true;
			/* Fallback if it's an embedded schema (no `path`) */
			const vFilePath = vFile.path || '';

			/* From current processed file directory  (e.g. `./foo…` or `../foo…`) */
			const dirFromCwd = path.isAbsolute(vFilePath)
				? path.relative(process.cwd(), path.dirname(vFilePath))
				: path.dirname(vFilePath);

			const standardPath = path.join(dirFromCwd, yamlJS.$schema);
			if (existsSync(standardPath)) {
				schemaPathFromCwd = standardPath;
			} else {
				/* Non standard behavior, like TS / Vite, not JSON Schema resolution.
          Resolving `/my/path` or `my/path` from current remark project root */
				schemaPathFromCwd = path.join(remarkCwd, yamlJS.$schema);
			}
		}
	} catch (error) {
		if (error instanceof Error) {
			const banner = `YAML frontmatter parsing: ${schemaPathFromCwd ?? ''}`;
			vFile.message(`${banner} — ${error.name}: ${error.message}`);
		}
	}

	/* ········································································ */

	/* Global schemas associations, only if no local schema is set */
	if (yamlDoc && yamlJS && !hasLocalAssoc) {
		Object.entries(settings.schemas ?? {}).forEach(
			([globSchemaPath, globSchemaAssocs]) => {
				/* Check if current markdown file is associated with this schema */
				globSchemaAssocs.forEach((mdFilePath) => {
					if (typeof mdFilePath === 'string') {
						const mdPathCleaned = path.normalize(mdFilePath);

						/* With `remark`, `vFile.path` is already relative to project root,
              while `eslint-plugin-mdx` gives an absolute path */
						const vFilePathRel = path.relative(remarkCwd, vFile.path);

						if (minimatch(vFilePathRel, mdPathCleaned)) {
							schemaPathFromCwd = path.join(remarkCwd, globSchemaPath);
						}
					}
				});
			},
		);
	}

	/* ········································································ */

	let schema: JSONSchema7 | undefined;
	if (hasPropSchema) {
		schema = settings.embed;
	} else if (schemaPathFromCwd) {
		/* Load schema + references */
		schema = await $RefParser
			// NOTE: Ext. `$refs` are embedded, not local defs.
			// Could be useful to embed ext. refs. in definitions,
			// so we could keep the ref. name for debugging?
			.bundle(schemaPathFromCwd)
			.catch((error) => {
				if (error instanceof Error) {
					const banner = `YAML schema file load/parse: ${
						schemaPathFromCwd ?? ''
					}`;
					vFile.message(`${banner} — ${error.name}: ${error.message}`);
				}
				return undefined;
			})
			/* Asserting then using a JSONSchema4 for AJV (JSONSchema7) is OK */
			.then((refSchema) =>
				refSchema ? (refSchema as JSONSchema7) : undefined,
			);

		/* Schema is now extracted,
      remove in-file `$schema` key, so it will not interfere later */

		if (hasLocalAssoc && yamlJS && typeof yamlJS.$schema === 'string') {
			delete yamlJS.$schema;
		}
	}

	/* ········································································ */

	/* We got an extracted schema to work with */
	if (schema && yamlDoc) {
		/* Setup AJV (Another JSON-Schema Validator) */
		const ajv = new Ajv({
			/* Defaults */
			allErrors: true /* So it doesn't stop at the first found error */,
			strict: false /* Prevents warnings for valid, but relaxed schemas */,

			/* User settings / overrides */
			...settings.ajvOptions,
		});
		addFormats(ajv);

		/* JSON Schema compilation + validation with AJV */
		try {
			const validate = ajv.compile(schema);
			validate(yamlJS);

			/* Push JSON Schema validation failures messages */
			if (validate.errors?.length) {
				pushErrors(
					validate.errors,
					yamlDoc,
					vFile,
					schemaPathFromCwd ?? '',
					lineCounter,
				);
			}
		} catch (error) {
			if (error instanceof Error) {
				const banner = `JSON schema malformed: ${schemaPathFromCwd ?? ''}`;
				vFile.message(`${banner} — ${error.name}: ${error.message}`);
			}
		}
	}
}

/* ·········································································· */

const remarkFrontmatterSchema = lintRule(
	{
		url,
		origin: 'remark-lint:frontmatter-schema',
	},
	async (ast: Root, vFile: VFile, settings: Settings = {}) => {
		if (ast.children.length) {
			/* Handle only if the processed Markdown file has a frontmatter section */
			const frontmatter = ast.children.find((child) => child.type === 'yaml');
			if (frontmatter?.type === 'yaml') {
				await validateFrontmatter(frontmatter, vFile, settings);
			}
		}
	},
);

export default remarkFrontmatterSchema;
